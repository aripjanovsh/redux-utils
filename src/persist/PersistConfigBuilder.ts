import { PersistConfig, Transform } from "redux-persist";

import { PlainObject } from "../core/types";
import {
  invariant,
  isFunction,
  isPlainObject,
  warning,
} from "../internal/utils";

export interface TransformConfig<S, R = any> {
  in: (state: S) => null | R;
  out: (raw: null | undefined | R) => null | S;
}

export class PersistConfigBuilder<S extends PlainObject> {
  private readonly initialState: S;

  private readonly whitelist: string[] = [];

  // eslint-disable-next-line typescript/no-explicit-any
  private readonly transforms: Array<Transform<any, any>> = [];

  public constructor(initialState: S) {
    invariant(
      isPlainObject(initialState),
      "ReducerBuilder: `initialState` expected to be a plain object.",
    );

    this.initialState = initialState;
  }

  public whitelistChildren(...childKeys: Array<keyof S>): this {
    childKeys.forEach(x => {
      const key = String(x);

      if (this.whitelist.indexOf(key) === -1) {
        this.whitelist.push(key);
      }
    });

    return this;
  }

  public addChildTransforms<K extends keyof S>(
    childKey: K,
    ...configs: Array<TransformConfig<S[K]>>
  ): this {
    invariant(
      childKey != null,
      "PersistConfigBuilder: `childKey` expected to be defined.",
    );

    const { [childKey]: childState } = this.initialState;

    this.whitelistChildren(childKey);

    invariant(
      configs.length > 0,
      "PersistConfigBuilder: at least one transform should be passed.",
    );

    configs.forEach(config => {
      invariant(
        config != null,
        "PersistConfigBuilder: `config` expected to be defined.",
      );

      const { in: inbound, out: outbound } = config;

      invariant(
        isFunction(inbound),
        "PersistConfigBuilder: `config.in` expected to be a function.",
      );

      invariant(
        isFunction(outbound),
        "PersistConfigBuilder: `config.out` expected to be a function.",
      );

      this.transforms.push({
        in: (x: S[K], k) => (k !== childKey ? x : inbound(x)),
        out: (x, k): S[K] => (k !== childKey ? x : outbound(x) || childState),
      });
    });

    return this;
  }

  public build(): Partial<PersistConfig> {
    warning(
      this.whitelist.length > 0,
      "PersistConfigBuilder: persist config has not whitelisted keys.",
    );

    return { whitelist: this.whitelist, transforms: this.transforms };
  }
}
