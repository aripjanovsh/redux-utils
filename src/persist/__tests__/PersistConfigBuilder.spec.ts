import { mockConsole } from "../../__tests__/mockConsole";
import { PersistConfigBuilder } from "../PersistConfigBuilder";

describe("PersistConfigBuilder", () => {
  const consoleError = jest.fn();

  mockConsole({ error: consoleError });

  test("init validation", () => {
    expect(() => {
      // @ts-ignore
      // eslint-disable-next-line no-new
      new PersistConfigBuilder([]);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      // eslint-disable-next-line no-new
      new PersistConfigBuilder(null);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      // eslint-disable-next-line no-new
      new PersistConfigBuilder(undefined);
    }).toThrowErrorMatchingSnapshot();
  });

  test("build without whitelisted keys", () => {
    const config = new PersistConfigBuilder({}).build();

    expect(config).toBeTruthy();

    expect(consoleError).toBeCalled();
    expect(consoleError.mock.calls).toMatchSnapshot();
  });

  test("duplicate whitelist keys", () => {
    const config = new PersistConfigBuilder({ foo: 1, bar: 2, baz: 3 })
      .whitelistChildren("foo", "bar")
      .whitelistChildren("bar", "baz")
      .whitelistChildren("baz", "foo")
      .build();

    expect(config.whitelist).toEqual(["foo", "bar", "baz"]);

    expect(consoleError).not.toBeCalled();
  });

  test("child transform validation", () => {
    const builder = new PersistConfigBuilder({ foo: 1 });

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms(null);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms(undefined);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms("foo");
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms("foo", null);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms("foo", undefined);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms("foo", {});
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms("foo", { in: () => {} });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      builder.addChildTransforms("foo", { out: () => {} });
    }).toThrowErrorMatchingSnapshot();

    expect(consoleError).not.toBeCalled();
  });

  test("child transform build", () => {
    const {
      whitelist,
      transforms,
      ...restConfigProps
    } = new PersistConfigBuilder({ foo: 1 })
      .addChildTransforms("foo", {
        in: x => JSON.stringify(x),
        out: x => JSON.parse(x),
      })
      .build();

    expect(transforms).toBeTruthy();
    expect(whitelist).toEqual(["foo"]);
    expect(restConfigProps).toEqual({});

    const [transform, ...restTransforms] = transforms!;

    expect(restTransforms).toEqual([]);

    const plain = { quoz: 1 };
    const json = JSON.stringify(plain);

    expect(transform.in(plain, "foo")).toBe(json);
    expect(transform.out(json, "foo")).toEqual(plain);

    expect(transform.in(plain, "bar")).toBe(plain);
    expect(transform.out(plain, "bar")).toBe(plain);

    expect(consoleError).not.toBeCalled();
  });

  test("empty child transform build", () => {
    const { whitelist, transforms } = new PersistConfigBuilder({ foo: 1 })
      .addChildTransforms("foo", {
        in: x => JSON.stringify(x),
        out: x => JSON.parse(x),
      })
      .build();

    expect(transforms).toBeTruthy();
    expect(whitelist).toEqual(["foo"]);

    const [transform] = transforms!;

    expect(transform.in(null, "foo")).toBe("null");
    expect(transform.out(null, "foo")).toBe(1);

    expect(consoleError).not.toBeCalled();
  });

  test("multiple child transforms build", () => {
    const { whitelist, transforms } = new PersistConfigBuilder({ foo: 1 })
      .addChildTransforms(
        "foo",
        { in: x => x * 2, out: x => x / 2 },
        { in: x => x + 10, out: x => x - 10 },
      )
      .build();

    expect(transforms).toBeTruthy();
    expect(whitelist).toEqual(["foo"]);

    const [t1, t2] = transforms!;

    expect(t1.in(1, "foo")).toBe(2);
    expect(t1.out(2, "foo")).toBe(1);

    expect(t2.in(1, "foo")).toBe(11);
    expect(t2.out(11, "foo")).toBe(1);

    expect(consoleError).not.toBeCalled();
  });
});
