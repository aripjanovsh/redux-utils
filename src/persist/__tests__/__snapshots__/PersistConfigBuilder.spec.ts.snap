// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersistConfigBuilder build without whitelisted keys 1`] = `
Array [
  Array [
    "PersistConfigBuilder: persist config has not whitelisted keys.",
  ],
]
`;

exports[`PersistConfigBuilder child transform validation 1`] = `"PersistConfigBuilder: \`childKey\` expected to be defined."`;

exports[`PersistConfigBuilder child transform validation 2`] = `"PersistConfigBuilder: \`childKey\` expected to be defined."`;

exports[`PersistConfigBuilder child transform validation 3`] = `"PersistConfigBuilder: at least one transform should be passed."`;

exports[`PersistConfigBuilder child transform validation 4`] = `"PersistConfigBuilder: \`config\` expected to be defined."`;

exports[`PersistConfigBuilder child transform validation 5`] = `"PersistConfigBuilder: \`config\` expected to be defined."`;

exports[`PersistConfigBuilder child transform validation 6`] = `"PersistConfigBuilder: \`config.in\` expected to be a function."`;

exports[`PersistConfigBuilder child transform validation 7`] = `"PersistConfigBuilder: \`config.out\` expected to be a function."`;

exports[`PersistConfigBuilder child transform validation 8`] = `"PersistConfigBuilder: \`config.in\` expected to be a function."`;

exports[`PersistConfigBuilder deprecated \`addChildTransform\` 1`] = `
Array [
  Array [
    "\`PersistConfigBuilder#addChildTransform\` is deprecated in favor of \`PersistConfigBuilder#addChildTransforms\`.",
  ],
]
`;

exports[`PersistConfigBuilder deprecated \`whitelistKeys\` 1`] = `
Array [
  Array [
    "\`PersistConfigBuilder#whitelistKeys\` is deprecated in favor of \`PersistConfigBuilder#whitelistChildren\`.",
  ],
]
`;

exports[`PersistConfigBuilder init validation 1`] = `"ReducerBuilder: \`initialState\` expected to be a plain object."`;

exports[`PersistConfigBuilder init validation 2`] = `"ReducerBuilder: \`initialState\` expected to be a plain object."`;

exports[`PersistConfigBuilder init validation 3`] = `"ReducerBuilder: \`initialState\` expected to be a plain object."`;
